ubuntu/entrypoint.sh

```diff
@@ -1,21 +1,48 @@
 #!/bin/bash -x
-mount_rootfs() {
+
+function exit_on_error() {
+    $@
+    if [[ $? -ne 0 ]]; then
+        echo "Error occured while executing: $1"
+        exit 1
+    fi
+}
+
+function mount_rootfs() {
     echo "Mounting OFED driver container rootfs..."
     mount --make-runbindable /sys
     mount --make-private /sys
     mkdir -p /run/mellanox/drivers
     mount --rbind / /run/mellanox/drivers
 }
 
-unmount_rootfs() {
+function unmount_rootfs() {
     echo "Unmounting Mellanox OFED driver rootfs..."
     if findmnt -r -o TARGET | grep "/run/mellanox/drivers" > /dev/null; then
       umount -l -R /run/mellanox/drivers
     fi
 }
 
-/etc/init.d/openibd restart
-ofed_info -s
+function rebuild_driver() {
+    # Rebuild driver in case installed driver kernel version differs from running kernel
+    echo "Rebuilding driver"
+    apt-get -yq update
+    apt-get -yq install linux-headers-$(uname -r)
+    dkms autoinstall
+}
+function start_driver() {
+    /etc/init.d/openibd stop
+    /etc/init.d/openibd start
```

Comment:
can't you do restart?

Reply:
i treat driver stop as best effort as it may fail if it tires to unload missing drivers.
and only look at the RC of driver start.

do you think a different approach is needed ?

Reply:
i think the missing drivers bit occurres when there is a kernel missmatch and the script fallsback to unloading everyting it knows, but this needs more investigation.

we could stick to just driver restart to be more strict.

Reply:
Thinking on this more it should be restart here.
thinking on a case where stop fails to unload when driver exists (e.g being used/ stuck)
then driver start will simply do nothing right ?

Reply:
so we need to stop in the background and and loop with a timeout and force kill with -9. or this is what the ofed stop is doing?

Reply:
i dont aim to deal with such cases, it may fail with "module being used" like you do rmmod/modprobe -r 

nothing should hang, and if it does its should be fixed at that level, not here.

Reply:
so let do restart